File: java/src/main/java/com/example/cote/PGMRS/체육복.java
package com.example.cote.PGMRS;

import java.util.*;
import java.io.*;
import java.nio.file.*;
import static java.util.stream.Collectors.*;

public class 체육복 {

	private static final int MAX_N = 30;

	private static int solve(int n, int[] lost, int[] reserve) {
		Arrays.sort(lost);
		Arrays.sort(reserve);

		// reserve 와 lost의 교집합
		Set<Integer> reserveAndLost = Arrays.stream(reserve)
			.boxed()
			.collect(toSet());
		reserveAndLost.retainAll(Arrays.stream(lost)
			.boxed()
			.collect(toSet()));

		Queue<Integer> q = new LinkedList();

		for (int l : lost) {
			if (reserveAndLost.contains(l)) {
				continue;
			}
			q.add(l);
		}

		int borrowers = 0;

		for (int r : reserve) {
			if (reserveAndLost.contains(r)) {
				continue;
			}

			while (!q.isEmpty() && q.peek() < r - 1 ){
				q.poll();
			}
			if (q.isEmpty()) break;

			if (q.peek() <= r+1) {
				q.poll();
				borrowers++;
			}
		}

		return n - lost.length + reserveAndLost.size() + borrowers;
	}

	public static void main(String... args) throws IOException {
		
		Path p = Paths.get(System.getProperty("user.dir") + "/data/체육복.txt");
		BufferedReader reader = Files.newBufferedReader(p);

		int testCases = Integer.parseInt(reader.readLine());

		while (testCases > 0) {
			int n = Integer.parseInt(reader.readLine());
			int[] lost = Arrays.stream(reader.readLine().split(" "))
				.mapToInt(Integer::parseInt)
				.toArray();
			int[] reserve = Arrays.stream(reader.readLine().split(" "))
				.mapToInt(Integer::parseInt)
				.toArray();
			int answer = Integer.parseInt(reader.readLine());

			System.out.printf("%d %d\n", answer, solve(n, lost, reserve));

			testCases--;
		}

	}
}

File: kotlin/src/main/kotlin/com/example/cote/PGMRS/체육복.kt
package com.example.cote.PGMRS

import java.io.*
import java.nio.file.*
import kotlin.random.*

object 체육복 {
    val MAX_N = 30
    val TEST_CASES = 10

    val random = Random(43)

    fun solve(n: Int, lost: List<Int>, reserve: List<Int>): Int {

        val lostAndReserve = lost.toSet().run { intersect(reserve.toSet()) }

        val lostAndNotReserve = lost.toSet().run { subtract(reserve.toSet()) }

        val helpee = ArrayDeque<Int>(lostAndNotReserve)

        var borrower = 0

        for (helper in reserve) {
            if (lostAndReserve.contains(helper)) {
                continue
            }

            while (!helpee.isEmpty() && helpee.first() < helper - 1) {
                helpee.removeFirst()
            }

            if (helpee.isEmpty()) {
                break
            }

            if (helpee.first() <= helper + 1) {
                helpee.removeFirst()
                borrower++
            }
        }

        return n - lost.size + borrower
    }

    fun main() {
        val reader =
            with(Paths.get(System.getProperty("user.dir") + "/data/체육복.txt")) {
                Files.newBufferedReader(this)
            }

        repeat(reader.readLine().toInt()) {
            val n = reader.readLine().toInt()
            val lost = reader.readLine().split(" ").map(String::toInt)
            val reserve = reader.readLine().split(" ").map(String::toInt)
            val answer = reader.readLine().toInt()

            println("$answer ${solve(n, lost, reserve)}")
        }

        repeat(TEST_CASES) {
            val n = random.nextInt(MAX_N - 2) + 2
            val lost = IntRange(1, n).filter { random.nextFloat() < 0.5 }
            val reserve = IntRange(1, n).filter { random.nextFloat() < 0.2 }
            println("$n ${solve(n, lost, reserve)}")
        }
    }
}

fun main() {
    체육복.main()
}
